password.c

```diff
@@ -78,28 +78,40 @@ static void schedule_password_clear(struct swaylock_state *state) {
 			state->eventloop, 10000, clear_password, state);
 }
 
+static struct swaylock_password *copy_password(struct swaylock_password pw) {
+	struct swaylock_password *copy = malloc(sizeof(pw));
+
+	copy->len = pw.len;
+	copy->buffer_len = pw.buffer_len;
+	copy->buffer = malloc(pw.len + 1);
+	strncpy(copy->buffer, pw.buffer, pw.len + 1); // strlcpy is not available?
+
+	return copy;
+}
+
 static void submit_password(struct swaylock_state *state) {
 	if (state->args.ignore_empty && state->password.len == 0) {
 		return;
 	}
 
 	state->auth_state = AUTH_STATE_VALIDATING;
 
-	if (!write_comm_request(&state->password)) {
+	// Create a copy to pass for validation.
+	struct swaylock_password *pw = copy_password(state->password);
+	// And free the buffer for anything that's typed during validation or later.
+	clear_password_buffer(&state->password);
+
+	if (!write_comm_request(pw)) {
```

Comment:
The write is non-blocking. Why do we need to copy?

Reply:
My bad, I was under the impression that we might alter the buffer while before it's written out, but this does actually block very briefly.

The updated approach works, but I'm noticing it allows trying a lot of attempts without any delay before write_comm_request between them. Even if pam complains `pam_authenticate failed: maximum number of authentication tries exceeded`, writing the right password without any pause after many failed tries still works. I think this would allow using a fake USB keyboard that just brute forces without limitation.